java/ql/lib/semmle/code/java/Expr.qll

```diff
@@ -1233,7 +1233,14 @@ class ClassInstanceExpr extends Expr, ConstructorCall, @classinstancexpr {
    * empty type argument list of the form `<>`.
    */
   predicate isDiamond() {
-    this.getType() instanceof ParameterizedClass and
+    (
+      this.getType() instanceof ParameterizedClass
+      or
+      this.getAnonymousClass().getASupertype() instanceof ParameterizedType and
+      // Ignore Kotlin code; otherwise seems to erroneously match `object` expression with explicit
+      // type arguments, possibly because extractor is not properly extracting the type args?
+      not this.getCompilationUnit().isKotlinSourceFile()
```

Comment:
@igfoo is there a quick fix for the Kotlin case rather than hack Kotlin source out like this? @Marcono1234 for clarity could you show a Kotlin snippet that otherwise behaves badly?

Reply:
If I omit that exclusion of Kotlin source files and then run the following query on https://github.com/square/okhttp
```ql
import java

from ClassInstanceExpr e
where
  e.isDiamond() and
  e.getCompilationUnit().isKotlinSourceFile()
select e
```

it finds for example this expression with an explicit type argument:
https://github.com/square/okhttp/blob/64d1262006c80a1eb3102ee7c647e8a840e0e54e/okhttp/src/jvmMain/kotlin/okhttp3/internal/http/dates.kt#L32C36-L32C70

Though maybe the `isDiamond` predicate does not make sense for Kotlin in the first place; at least there is (to my knowledge) no `<>` syntax but instead the type arguments are simply omitted if they can be inferred by the compiler.
Or at least the name `isDiamond` might be a bit confusing then.

Reply:
Yep, that looks like a bug in our Kotlin support at the moment. Excluding like this looks fine for now. I've made an internal ticket to remind us to come back to this.